### PR TITLE
Improve listing performance of Hudi tables

### PR DESCRIPTION
**Pull request by codope**
- Integrate metadata-based listing for Hudi tables. This is
enabled by a session property.
- Implement a new DirectoryLister for Hudi that uses
HoodieTableFileSystemView to fetch data files.
- Bump Hudi version to 0.10.1 to use above features.
- Replace hudi-common, hudi-hadoop-mr by
hudi-presto-bundle.

Test plan - (Please fill in how you tested your changes)

- Synced a Hudi table to Hive and queried through Presto.
- Add unit test for HudiDirectoryLister.

```
== RELEASE NOTES ==

Hive Changes
* Upgrade Hudi version to 0.10.1.
* Support metadata-based listing and bootstrap for Hudi tables.
```

### CODE REVIEW

**Comment by codope** [comment]
@arunthirupathi this is rework of #17084 with thinned presto bundle. Could you please review?
cc @vinothchandar

**Comment by arunthirupathi** [comment]
Can you post the mvn dependency difference before this change and after this change ?

What is the size of the shaded jar ? 

I made a pass on this PR and it looks good. I need to sync up to this PR and do an internal build to ensure it does not regress things.

**Comment by arunthirupathi** [comment]
I will update by this Wednesday on whether this PR works with our internal dependencies correctly.

File: pom.xml [hunk 1]
```diff
@@ -1189,30 +1189,15 @@
...
-                        <artifactId>objenesis</artifactId>
-                    </exclusion>
-                    <exclusion>
-                        <groupId>commons-logging</groupId>
-                        <artifactId>commons-logging</artifactId>
-                    </exclusion>
-                    <exclusion>
-                        <groupId>org.slf4j</groupId>
-                        <artifactId>jcl-over-slf4j</artifactId>
+                        <groupId>com.esotericsoftware</groupId>
```
Comment (arunthirupathi) [inline pom.xml:1199]: Why are their still excludes if the hudi-presto-bundle is shaded ?

Comment (codope) [reply pom.xml]: Removed unnecessary exclusions. There are some hudi modules which we need to exclude and then there is protobuf, which if not excluded, was causing the build to fail due to some conflicting resource files.

**Comment by codope** [comment]
@arunthirupathi Could you please take a look again? I have removed unnecessary excludes. Also, was there any issue you fuond out running this internally?

**Comment by arunthirupathi** [comment]
There was a minor change required due to PathFilter to Optional<PathFilter>, but the build passed. I am going to kick off integration tests to see if it causes any issues.

**Comment by arunthirupathi** [comment]
the integration tests succeeded, so this is good. I will take  final pass on Tuesday and merge this in.

**Comment by codope** [comment]
> the integration tests succeeded, so this is good. I will take final pass on Tuesday and merge this in.

Glad to hear this. Thanks for the update.

**Comment by arunthirupathi** [comment]
Can you please rebase and push ? There is a breaking change in our internal repo, I have submitted a PR for it. Once that is approved, I will merge this change and the other change together.

**Review by arunthirupathi — APPROVED** [review]

File: presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java [hunk 2]
```diff
@@ -31,6 +31,7 @@
 import java.util.Collections;
 import java.util.Deque;
 import java.util.Iterator;
+import java.util.Optional;
```
Comment (arunthirupathi) [inline presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java:34]: On a second thought it will be easier, if you can undo this change to this file.

Where you need the filter to always work, pass in path filter where it always returns true. () -> true

This will make it easier for me to merge this PR.

Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java:34]: PathFilter was added to DirectoryLister [due to Hudi](https://github.com/prestodb/presto/pull/13818/files#diff-d8caebd0319edf28ebbe49a5370a8a59f5bc767726f321ceab12e88ead8a2fc3). After this patch, we don't really need it. I was planning to remove that as a follow-up to this PR if that's ok with you?

File: presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java [hunk 3]
```diff
@@ -259,7 +260,7 @@ public StoragePartitionLoader(
 
             return addSplitsToSource(splits, splitFactory, hiveSplitSource, stopped);
         }
-        PathFilter pathFilter = isHudiParquetInputFormat(inputFormat) ? hoodiePathFilterLoadingCache.getUnchecked(configuration) : path1 -> true;
+        PathFilter pathFilter = path1 -> true;
```
[the remark this answers is not in the document]
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java:263]: Please see my comment below.

File: presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java (hunk 2 above)
Comment (arunthirupathi) [reply presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java:34]: Let us do it with this PR itself.

Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java:34]: got it.. will update this PR in a day or two.

Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java]: updated the PR

File: presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+public class HudiDirectoryLister
+        implements DirectoryLister
+{
+    private static final Logger log = Logger.get(HudiDirectoryLister.class);
+
+    private final HoodieTableFileSystemView fileSystemView;
+    private final HoodieTableMetaClient metaClient;
+    private final boolean metadataEnabled;
+
+    public HudiDirectoryLister(Configuration conf, ConnectorSession session, Table table)
```
[the remark this answers is not in the document]
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:53]: That's more of a legacy thing. For the outside world it is Hudi. 
Hudi started as `Hoodie` but later renamed to `Hudi`, however we did not change the class names.

**Comment by codope** [comment]
@arunthirupathi This PR is ready now. I have removed the path filter form DirectoryLister interface and tested locally.

**Comment by arunthirupathi** [comment]
This change looks good, since this breaks API, I need to have a PR for our internal changes get it approved and then merge this in at the same time.

**Comment by codope** [comment]
Thanks @arunthirupathi for shepherding and landing this!
